Expand class-level documentation for `WP_HTML_Tag_Processor`

dmsnell opened the pull request:
## What

Also fixes `get_tag()` to return upper-case variant of tag names in accordance with the inline documentation and with other DOM interfaces.

## Why?

Because we want to help people understand how to use it, and get that information at the cursor where they're typing.

## How?

Adds documentation in docblock comments.

## Testing Instructions

The only thing to test is that `get_tag()` returns an upper-case tag name, which should be possible by inspection.

dmsnell committed:
Expand class-level documentation for `WP_HTML_Tag_Processor`

Also fixes `get_tag()` to return upper-case variant of tag names in
accordance with the inline documentation and with other DOM interfaces.

adamziel commented on lib/experimental/html/class-wp-html-tag-processor.php:
````diff
@@ -29,6 +32,144 @@
+ * Example
+ * ```php
+ *     $processor = new WP_HTML_Tag_Processor( $html );
+ *     if ( $processor->next_tag( array( 'tag_name' => 'option' ) ) ) {
````
What do you think about surfacing the shorthand form?

```suggestion
 *     if ( $processor->next_tag( 'option' ) ) {
```

dmsnell replied:
definitely not in this PR, but overall I'm not sure.

on one hand I think it's convenient, but on the other it does muddy the consistency, and I think it's likely more common that people will search by class anyway rather than tag, or at least reasonably as often as by class.

it did occur to me that maybe `tag` and `class` would suffice, instead of `tag_name` and `class_name`

on this note I wonder if we have to stick to the long array syntax in our example code. we have to use those in the code for PHP support, but are we required to demonstrate as examples the use of the more restricted syntax?

adamziel replied:
> I think it's likely more common that people will search by class anyway rather than tag, or at least reasonably as often as by class.

I'd say 50/50 is as good of a guess as we can make right now.

> it did occur to me that maybe tag and class would suffice, instead of tag_name and class_name

Yes please!

>  are we required to demonstrate as examples the use of the more restricted syntax?

That is a good question, I don't know! @costdev what do you think?

costdev replied:
There's ~11 examples of short array syntax within docblocks/inline comments, and ~53 examples of long array syntax. While [this example](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#description) (see point 2) shows the long syntax, the accompanying text doesn't have anything to do with the "right" syntax to use. I'd suggest just going with best judgement on readability.

However, pinging @SergeyBiryukov as someone who has probably seen more Core docblocks than anyone.

dmsnell replied:
I've replaced the examples with shorthand array syntax for brevity and clarity. We can remove them later if we decide that it's best to leave in the `array()` notation.

31402ddfc8

adamziel commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -29,6 +32,144 @@
+ * | Goal                                                      | Query                                                                              |
+ * |-----------------------------------------------------------|------------------------------------------------------------------------------------|
+ * | Find any tag.                                             | `$processor->next_tag();`                                                          |
+ * | Find next image tag.                                      | `$processor->next_tag( array( 'tag_name' => 'img' ) );`                            |
```
Let's mention the shorthand form here:

```suggestion
  * | Find next image tag.                                      | `$processor->next_tag( 'img' );`                                                   |
  * | Find next image tag.                                      | `$processor->next_tag( array( 'tag_name' => 'img' ) );`                            |
```

dmsnell replied:
see my previous comment on this. I have second-guesses about this short-hand

adamziel replied:
Wait, do you mean second thoughts on surfacing it here, or second thoughts on even supporting it?

dmsnell replied:
supporting it at all. my worry is that we say you can search for a given tag name, but then people want a short hand for class name, so we add it, but then if you don't specify a tag you have `->next( null, 'wp-block-group' )` everywhere. or someone things, let's put class name first, and then we have the other problem.

I'd even be for exposing this shorthand only after we see use in practice to see if people even want it, and if so, which convention is more dominant.

alternatively it's always easy to later add new specialized methods. suppose we only expose `->next()` right now, taking a query object. we can add `->next_tag( 'img' )` and `->next_with_class( 'wp-block-group' )`. performance wise this is a "free" option

adamziel replied:
> I'd even be for exposing this shorthand only after we see use in practice to see if people even want it, and if so, which convention is more dominant.

Let's do that 👍  Renaming this method to just `next( $query )` also works for me as `next_tag( $tag_name )` makes a lot of sense.

adamziel commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -29,6 +32,144 @@
+ *              'jazzy' === $processor->get_attribute( 'data-style' )
+ *         ) {
+ *             $processor->add_class( 'theme-style-everest-jazz' );
+ *             if ( 0 === --$remaining_tag_to_style ) {
```
I missed the `--` the first time I read this, let's split it into two lines to make the intention clearer:

```suggestion
 *             $remaining_tag_to_style -= 1;
 *             if ( 0 === $remaining_tag_to_style ) {
```

dmsnell replied:
resolved in be5015642e by moving the condition into the `while`

adamziel commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -29,6 +32,144 @@
+ *
+ * ### Modifying HTML attributes for a found tag
+ *
+ * Once the processor has found the start of a tag you can
```
```suggestion
 * Once the processor has found the start of a tag, you can
```

dmsnell replied:
for this and the other comma suggestions I believe they are [unnecessary](https://owl.excelsior.edu/grammar-essentials/common-errors/common-errors-unnecessary-comma/), so I'd rather leave them out.

adamziel commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -29,6 +32,144 @@
+ * ### Modifying HTML attributes for a found tag
+ *
+ * Once the processor has found the start of a tag you can
+ * instruct it to modify the attributes on that tag by setting
+ * a new value for an attribute or by removing an existing one.
```
```suggestion
 * create, update, and remove the attributes of that tag.
```

dmsnell replied:
I intentionally worded it to match the `set` and `remove` terminology of the API (and thus suggest how to use it), but this reads more casually. what do you think about this? we don't currently have any functions named to create or update attribute values…

adamziel replied:
hm yeah I think I agree. How about this, then?

```suggestion
 * set a new value of tag attributes or remove them entirely.
```

dmsnell replied:
reworded in e3f84affec - happy to continue to adjust if it's not ideal

> Once you've found the start of an opening tag you can modify
> any number of the attributes on that tag. You can set a new
> value for an attribute, remove the entire attribute, or do
> nothing and move on to the next opening tag.

adamziel replied:
It's perfect, thank you

adamziel commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -29,6 +32,144 @@
+ * instruct it to modify the attributes on that tag by setting
+ * a new value for an attribute or by removing an existing one.
+ *
+ * Example
```
```suggestion
 * Example:
```

adamziel commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -29,6 +32,144 @@
+ * inspect the search results using the read-only functions
+ * provided by the processor or external state or variables.
+ *
+ * Example
```
```suggestion
 * Example:
```

adamziel commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -29,6 +32,144 @@
+ *  2. Find the tag(s) you are looking for.
+ *  3. Request changes to the attributes in those tag(s).
+ *
+ * Example
```
```suggestion
 * Example:
```

dmsnell replied:
I'm going to merge these in, but I also question if they should be here. In my persistent quest to understand docblock rendering I think at one point I even had an issue with `Example:` that `Example` didn't have.

adamziel commented on lib/experimental/html/class-wp-html-tag-processor.php:
````diff
@@ -29,6 +32,144 @@
+ * Example
+ * ```php
+ *     if ( $processor->next_tag( array( 'class' => 'wp-group-block' ) ) ) {
+ *         $processor->set_attribute( 'title' => 'This groups the contained content.' );
````
```suggestion
 *         $processor->set_attribute( 'title', 'This groups the contained content.' );
```

dmsnell replied:
very nice catch, thank you!

adamziel commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -29,6 +32,144 @@
+ * ### Modifying CSS classes for a found tag
+ *
+ * The tag processor treats the `class` attribute as a special case.
+ * Because it's a common operation to add or remove CSS classes you
```
```suggestion
 * Because it's a common operation to add or remove CSS classes, you
```

adamziel commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -29,6 +32,144 @@
+ *
+ * As with attribute values, adding or removing CSS classes is a safe
+ * operation that doesn't require checking if the attribute or class
+ * exists before making changes. If removing the only class then the
```
```suggestion
 * exists before making changes. If removing the only class name, the
```

adamziel commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -29,6 +32,144 @@
+ * exists before making changes. If removing the only class then the
+ * entire `class` attribute will be removed.
+ *
+ * Example
```
```suggestion
 * Example:
```

adamziel commented:
This is such a great documentation @dmsnell! 🎉  💯  🎉 

I left some tactical notes, but overall this is great – thank you so much! ✅

adamziel commented on lib/experimental/html/class-wp-html-tag-processor.php:
```diff
@@ -18,6 +18,9 @@
  * @TODO: Properly escape attribute value in `set_attribute()`
+ * @TODO: Add slow mode to escape character entities in CSS class names?
+ *        (This requires a custom decoder since `html_entity_decode()`
+ *        doesn't handle attribute character reference decoding rules.
```
Do you have an example? A it seems to decode sequences like `&#x042;` or `&#x2029;`

adamziel replied:
Gotcha:
```
php > echo html_entity_decode('&#x042');
&#x042
```

dmsnell replied:
I think something is wrong with that snippet as the output for that line should be `B` (and it is for me in a PHP shell).

to borrow an example from the HTML spec:

> If the markup contains (not in an attribute) the string `I'm &notit; I tell you`, the character reference is parsed as "not", as in, `I'm ¬it; I tell you` (and this is a parse error). But if the markup was `I'm &notin; I tell you`, the character reference would be parsed as "notin;", resulting in `I'm ∉ I tell you` (and no parse error).
>
> However, if the markup contains the string `I'm &notit; I tell you` in an attribute, no character reference is parsed and string remains intact (and there is no parse error).

![Screen Shot 2022-09-27 at 12 13 27 PM](https://user-images.githubusercontent.com/5431237/192615423-e013b851-940b-4adc-a735-7d3fb63fd74c.png)

And this is where PHP will mistranslate because it doesn't offer a flag (at least in any function I've found) which indicates the context in which a named character reference is being decoded.

adamziel replied:
> I think something is wrong with that snippet as the output for that line should be B (and it is for me in a PHP shell).

Interesting! I checked on PHP 8.1 and 7.4 and both outputted `&#042`.

Either way your argument makes sense :+1:

dmsnell replied:
> Interesting! I checked on PHP 8.1 and 7.4 and both outputted `&#042`.

In review I realize I inserted the missing semicolon in your example code, not having noticed it was missing. After fixing that I can confirm I'm seeing the same behavior your are, and also that the browser adds the missing semicolon (which means it diverges from how PHP handles it, and how the HTML spec suggests it ought to be handled).

<img width="310" alt="Screen Shot 2022-09-28 at 10 32 04 AM" src="https://user-images.githubusercontent.com/5431237/192850486-cefa76b1-a650-4ef3-8f53-ef79c76c0d33.png">

In ambiguous cases for obvious reasons it continues to consume the digits if it can.

<img width="312" alt="Screen Shot 2022-09-28 at 10 38 05 AM" src="https://user-images.githubusercontent.com/5431237/192851194-04c0fa2d-23f3-469e-87be-9173ce690cea.png">

And if the number exceeds the Unicode code point range it does what we expect from the spec and replaces it with a replacement symbol.

<img width="366" alt="Screen Shot 2022-09-28 at 10 38 18 AM" src="https://user-images.githubusercontent.com/5431237/192851262-2480bb1d-3e40-448f-a78f-73680e533a6b.png">

dmsnell committed:
Apply suggestions from code review

Author: @adamziel

Co-authored-by: Adam Zielinski <adam@adamziel.com>

dmsnell committed:
Rework loop counter in example

dmsnell commented:
Rebased from be50156 to 9a50202 with merge conflicts from `trunk`

dmsnell committed:
Use array shorthand notation in example docs

dmsnell committed:
Fix upper-case/lower-case confusion with tag names.

For consistency with other DOM libraries we're using upper-case tag names,
but we hadn't updated the code to reflect that.

dmsnell committed:
Reference $tags-> instead of $processor->

dmsnell committed:
undo rename
